Add technical steering proposal for `Spegel` registry extension (#814)
gardener/documentation: website/community/steering/technical/2026-02-spegel-registry-extension.md

- [Motivation](#motivation)
- [Proposal](#proposal)
  - [Introduction](#introduction)
  - [Registry Spegel API Design](#registry-spegel-api-design)
  - [Architecture Overview](#architecture-overview)
  - [Spegel Binary and Systemd Units](#spegel-binary-and-systemd-units)
  - [Containerd Configuration](#containerd-configuration)
  - [Spegel Bootstrapper](#spegel-bootstrapper)
  - [Observability and Monitoring](#observability-and-monitoring)
  - [Future Enhancement](#future-enhancement)
- [Impact and Alternatives](#impact-and-alternatives)
- [Decision Request](#decision-request)
- [Appendix (Optional)](#appendix-optional)

## Motivation

Current solution for image cache is based on [distribution/distribution](https://github.com/distribution/distribution) pull-through cache.

### The problem

The adoption of the registry cache extension is around 10% and not all used upstreams are covered by the extension's configurations. This is likely due to:
- **Complex configuration**: A separate [configuration](https://github.com/gardener/gardener-extension-registry-cache/blob/main/docs/usage/registry-cache/configuration.md#shoot-configuration) needs to be set for each upstream registry.

> - **Additional resources**: A StatefulSet is created for each of the configured upstream. The PV size (defaults to 10Gi) should be monitored and [increased](https://github.com/gardener/gardener-extension-registry-cache/blob/main/docs/usage/registry-cache/configuration.md#increase-the-cache-disk-size) if necessary, otherwise the cache becomes ineffective.

Reviewer: The link can point to a concrete tag. The content in the main branch may change over (for example to be moved to another location) time which will make the link a broken link.

```suggestion
- **Additional resources**: A StatefulSet is created for each of the configured upstream. The PV size (defaults to 10Gi) should be monitored and [increased](https://github.com/gardener/gardener-extension-registry-cache/blob/v0.19.0/docs/usage/registry-cache/configuration.md#increase-the-cache-disk-size) if necessary, otherwise the cache becomes ineffective.
```

Can you update everywhere to point to the latest tag (v0.19.0) instead of the master branch?

  - [TTL](https://github.com/gardener/gardener-extension-registry-cache/blob/main/docs/usage/registry-cache/configuration.md#garbage-collection) garbage collection setting is inefficient because it is not extended on access. Hence, an image is always deleted from the cache when its TTL expires, and will be re-pulled from upstream if requested again.

### Why Should We Care

Due to the low adoption rate and incomplete coverage of upstream registers, there are significant NAT Gateways traffic costs for downloading images.
Additionally, there are costs for outbound traffic from image registries like `Amazon Elastic Container Registry`, `Google Cloud Artifact Registry` and `Azure Container Registry`.
Even if the pull-through cache is configured, there are also costs for cross zonal traffic.

Providing a p2p image caching solution that is enabled by default and spans all upstream registries will reduce cloud costs and reduce latency when pulling images.

> ### Who benefits

Reviewer 2: ```suggestion
### Who Benefits
```

- **End users** benefit from the reduced latency when pulling images.
- **Stakeholders** see reduced costs for NAT Gateways and registries.
- **Gardener operators** are no longer engaged in monitoring notifications of PV sizes in the registry caches.

## Proposal

> The proposal is to extend the existing [registry-cache](https://github.com/gardener/gardener-extension-registry-cache) extension with a p2p image cache based on Spegel. New extension type `resistry-spegel` will be enabled by default for all Shoot clusters and will be used as [`_default`][setup-default-mirror-for-all-registries] mirror configuration in `containerd`.

Reviewer 2: ```suggestion
The proposal is to extend the existing [registry-cache](https://github.com/gardener/gardener-extension-registry-cache) extension with a p2p image cache based on Spegel. A new extension type `registry-spegel` will be enabled by default for all Shoot clusters and will be used as the [`_default`][setup-default-mirror-for-all-registries] mirror configuration in `containerd`.
```

### Introduction

A brief overview of the components used in this proposal.

#### Spegel overview
[Spegel][spegel] is a p2p image cache. It runs a local image registry on each node in a Kubernetes cluster. The local registry subscribes for containerd's image related events and serves image blobs and manifests that are available in the containerd image store.

> Content discovery in Kubernetes cluster is based on [Kademlia DHT][content-provider-routing]. When an image content (blob, manifest or index) is available in the containerd image store, Spegel adds its `digest` to the DHT provider store, announcing that the node provides the content corresponding to the `digest`. Then, when the same `digest` is needed by another node, Spegel searches the DHT for peers that provide the content and pull it from them.

Reviewer 2: ```suggestion
Content discovery in Kubernetes cluster is based on [Kademlia DHT][content-provider-routing]. When an image content (blob, manifest or index) is available in the containerd image store, Spegel adds its `digest` to the DHT provider store, announcing that the node provides the content corresponding to the `digest`. Then, when the same `digest` is needed by another node, Spegel searches the DHT for peers that provide the content and pulls it from them.
```

> Content discovery in Kubernetes cluster is based on [Kademlia DHT][content-provider-routing]. When an image content (blob, manifest or index) is available in the containerd image store, Spegel adds its `digest` to the DHT provider store, announcing that the node provides the content corresponding to the `digest`. Then, when the same `digest` is needed by another node, Spegel searches the DHT for peers that provide the content and pull it from them.

Reviewer: ```suggestion
The content discovery in a Kubernetes cluster is based on [Kademlia DHT][content-provider-routing]. When an image content (blob, manifest or index) is available in the containerd image store, Spegel adds its `digest` to the DHT provider store, announcing that the node provides the content corresponding to the `digest`. Then, when the same `digest` is needed by another node, Spegel searches the DHT for peers that provide the content and pull it from them.
```

In the Shoot control plane a Spegel `bootstrapper` is provided. It consist of:
- deployment
- service
- ingress

The `spegel` registries send a GET request to `https://<ingress-host>/bootstrap-nodes` to get the bootstrap peers. Traffic is encrypted using mTLS.
The `bootstrapper` uses client-go to access the `kube-apiserver` and list the Kubernetes nodes. It sorts the nodes and returns a subset of the first few node `net.IPAddr` addresses, similar to what is done in [`DNSBootstrapper`](https://github.com/spegel-org/spegel/blob/main/pkg/routing/bootstrap.go#L77-L80).

### Observability and Monitoring

The `spegel` registry exposes the following noticeable metrics:
- `http_response_size_bytes` - The size of the HTTP responses - histogram type.
- `http_request_duration_seconds` - The latency of the HTTP requests - histogram type.
- `spegel_mirror_requests_total` - Total number of mirror requests - counter type.
- `spegel_resolve_duration_seconds` - The duration for router to resolve a peer - histogram type.
- `spegel_advertised_keys` - Number of keys advertised to be available - gauge type.

With these metrics Spegel cache efficiency can be tracked.

### Future Enhancement

> In the [Network topology awareness][network-topology-awareness] issue there are options to implement zone aware routing (e.g. enhance Kademlia DHT with Soft Partitioning).

Reviewer: ```suggestion
### Future Enhancement

In the [Network topology awareness][network-topology-awareness] issue there are options to implement zone aware routing (e.g. enhance Kademlia DHT with Soft Partitioning).
```

A simple approach was tested locally on GCP infrastructure. When the content is available on the node, it registers `<digest>` and `<zone><digest>` as keys in the DHT. The disadvantage of this approach, besides duplication of keys, is that when searching for content, the `<zone><digest>` key is tried first, and if nothing is found, a new search is performed for the `<digest>` key.

### Proof of Concept
There is **PoC** of the above proposal available here:
- https://github.com/dimitar-kostadinov/gardener-extension-registry-cache/tree/spegel_api_poc
- https://github.com/dimitar-kostadinov/spegel/tree/poc

## Impact and Alternatives

### Risks, Downsides and Trade-offs
The external `Bootstrapper` should be contributed to Spegel.

### Alternative approaches

> [`Dragonfly`][dragonfly] was evaluated as an alternative. Unlike Spegel, it is not stateless and requires central components such as:

Reviewer: You could add empty line:
```suggestion
### Alternative approaches

[`Dragonfly`][dragonfly] was evaluated as an alternative. Unlike Spegel, it is not stateless and requires central components such as:
```
